mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionAvailbilityAwareAcivity.java

```diff
@@ -0,0 +1,74 @@
+package org.openhab.habdroid.core.connection;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.os.Bundle;
+import android.support.annotation.Nullable;
+import android.support.v4.app.TaskStackBuilder;
+import android.support.v7.app.AppCompatActivity;
+
+import org.openhab.habdroid.core.connection.exception.NetworkNotAvailableException;
+import org.openhab.habdroid.core.connection.exception.NetworkNotSupportedException;
+import org.openhab.habdroid.core.connection.exception.NoUrlInformationException;
+import org.openhab.habdroid.ui.NoNetworkActivity;
+import org.openhab.habdroid.ui.OpenHABPreferencesActivity;
+
+import static org.openhab.habdroid.ui.NoNetworkActivity.NO_NETWORK_MESSAGE;
+import static org.openhab.habdroid.ui.OpenHABPreferencesActivity.NO_URL_INFO_EXCEPTION_EXTRA;
+import static org.openhab.habdroid.ui.OpenHABPreferencesActivity.NO_URL_INFO_EXCEPTION_MESSAGE;
+
+public abstract class ConnectionAvailbilityAwareAcivity extends AppCompatActivity {
+    private Thread.UncaughtExceptionHandler originalHandler;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setConnectionExceptionHandler();
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+
+        if (originalHandler == null) {
+            return;
+        }
+
+        Thread.setDefaultUncaughtExceptionHandler(originalHandler);
+    }
+
+    private void setConnectionExceptionHandler() {
+        final Activity activity = this;
+
+        originalHandler = Thread.getDefaultUncaughtExceptionHandler();
+
+        Thread.setDefaultUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
```

Comment:
Replacing the uncaught exception handler for the main thread seems a bit heavy handed (and dirty). I'd prefer to make this logic more explicit by catching the relevant exceptions in the HTTP client code and routing it into the activity.

Reply:
The exceptions aren't thrown in the HTTP client, they're thrown during the try to construct a Connection object, and only when using Connections.ANY. This handler also does not handle all possible connection errors that may occur, such as failures when requesting a http reqsource or something like that. These kind of exceptions, if some are thrown, will still be handed over to the calling code an can be handled there (and if not handled at all, the default uncaught exception handler will step in).

This code also is meant for a really highlevel handling, so not every time someone is calling COnnectionFactory.getConnection(Connections.ANY, ctx); needs to handle the three possible exceptions. Otherwise, the app may only crash under certain conditions, which wouldn't be very good. In cases, where a specific action should be taken, such as showing a specific dialog to the user or whatever, the calling code should (still) handle the exceptions by themself.

Reply:
Why not give this class a getConnection() method that does the exception catching + handling? If the intention is that non-activity code that runs while the activity is active also gets the exception handling behavior, I'd oppose this even more: Such 'magical' behavior makes it hard to follow the code flow by code inspection and would better be avoided by having a convenience method for the exception handling that explicitly passes the exception to the activity. I did something like this in the past: https://github.com/slapperwan/gh4a/blob/master/app/src/main/java/com/gh4a/BaseActivity.java#L175

Reply:
I've now changed it in the way you described. I thought there was calls to the ConnectionFactory without an Activity, but it turns out there isn't. So the exception handling can be made in a getConnection() method.